src/metabase/feature_extraction/automagic_dashboards.clj

```diff
@@ -0,0 +1,335 @@
+    (events/publish-event! :dashboard-create dashboard)
+    dashboard))
+
+(defn- create-collection!
+  [name color description]
+  (when api/*is-superuser?*
+    (db/insert! 'Collection
+      :name        name
+      :color       color
+      :description description)))
+
+(def automagic-collection (delay
+                           (or (db/select-one 'Collection
+                                 :name "Automagic dashboards cards")
+                               (create-collection! "Automagic dashboards cards"
+                                                   "#000000"
+                                                   "All the cards we have automatically created for you."))))
+
+(defn- create-card!
+  [{:keys [visualization title description query]}]
+  (let [[visualization visualization-settings] (if (sequential? visualization)
+                                                 visualization
+                                                 [visualization {}])
+        card (db/insert! Card
+               :creator_id             api/*current-user-id*
+               :dataset_query          query
+               :description            description
+               :display                (or visualization :table)
+               :name                   title
+               :visualization_settings visualization-settings
+
+               :result_metadata        (card.api/result-metadata-for-query query)
+               :collection_id          (-> automagic-collection deref :id))]
+    (events/publish-event! :card-create card)
+    (hydrate card :creator :dashboard_count :labels :can_write :collection)
+    card))
+
+(defn- add-to-dashboard!
+  [dashboard card]
+  (dashboard/add-dashcard! dashboard (create-card! card)
+    (merge (next-card-position dashboard)
+           {:sizeX card-width
+            :sizeY card-height})))
+
+(defn- ensure-seq
+  [x]
+  (if (sequential? x)
+    x
+    [x]))
+
+(defn- build-query
+  [database table-id filters metrics dimensions limit order_by]
+  (let [query {:type     :query
+               :database database
+               :query    (cond-> {:source_table table-id}
+                           (not-empty filters)
+                           (assoc :filter (apply vector :and filters))
+
+                           (not-empty dimensions)
+                           (assoc :breakout dimensions)
+
+                           metrics
+                           (assoc :aggregation metrics)
+
+                           limit
+                           (assoc :limit limit))}]
+    (when (perms/set-has-full-permissions-for-set?
+           @api/*current-user-permissions-set*
+           (card/query-perms-set query :write))
+      query)))
+
+(defn- card-candidates
+  [context {:keys [metrics filters dimensions score limit order_by] :as card}]
+  (let [filters    (some->> filters
+                            ensure-seq
+                            (map (partial get (:filters context))))
```

Comment:
passing around context here is too cute. you should just destructure upstream and not require the card-candidates function to be aware of the structure of the yaml file

Reply:
Context has nothing to do with the yaml file (at least not directly) but holds all the bindings, db, and table(s). I think it's quite sensible organization: take a card definition and produce all possible cards given context.

It would be cleaner to defer creating the actual query, but then we run the risk of it being a dud if you don't have permissions (although that would require a pretty esoteric permissions, so maybe we can discount it).

Reply:
you're correct. I mis-read where context was coming from.